### PR TITLE
Go 1.6

### PR DESCRIPTION
**Pull request by ricochet1k**
Store all the folds in a map and pass an atomically incremented key instead of a Go pointer. This works well, but currently, nothing can ever be garbage collected once it passes through the map. This is obviously a problem and is going to need to be fixed soon. Ideas?

### CODE REVIEW

**Comment by erbridge** [comment]
On first run of lime-qml built with this and the `go16` branch of `gopy`, I get this panic:

```
Panic in RunMain func: runtime error: hash of unhashable type func()
fatal error: concurrent map writes

1: running [locked]
    debug   stack.go:24                 Stack(0, 0, 0)
    debug   stack.go:16                 PrintStack()
    qml-go  bridge.go:253               hookIdleTimer.func1.1()
            panic.go:443                panic(interface{}(0x861960))
    qml-go  qmlcommon.go:444            (*Common).On.func1()
    qml-go  bridge.go:110               RunMain(func(0xc820430960))
    qml-go  qmlcommon.go:449            (*Common).On(*Common(#3), string(0x8fb950, len=9), interface{}(0x7bad20))
    qml-go  qmlcommon.go:48             (*Common).setAddr(*Common(#3), Pointer(0x1b26a70))
    qml-go  qmlengine.go:45             NewEngine.func1()
    qml-go  bridge.go:257               hookIdleTimer.func1(#2)
    qml-go  bridge.go:258               hookIdleTimer()
    qml-go  ??:0                        _cgoexpwrap_838a2abd9f94_hookIdleTimer()
    qml-go  ??:0                        _Cfunc_applicationExec()
    qml-go  bridge.go:98                RunArgs([]string(#1 len=1 cap=1), func(#4), 0x0)
    qml-go  bridge.go:58                Run(func(#4), 0x0)
    main    frontend.go:63              initFrontend()
    main    main.go:34                  main()
1: running [locked]
    runtime panic.go:547                throw(0x93a610, 0x15)
    runtime hashmap.go:445              mapassign1(*maptype(0x7b6a20), *hmap(#10), Pointer(#20))
    qml-go  qmlcommon.go:444            (*Common).On.func1()
    qml-go  bridge.go:257               hookIdleTimer.func1(#2)
    qml-go  bridge.go:258               hookIdleTimer()
    qml-go  ??:0                        _cgoexpwrap_838a2abd9f94_hookIdleTimer()
    runtime asm_amd64.s:472             call32(0, #37, #38, 0)
    runtime cgocall.go:267              cgocallbackg1()
    runtime cgocall.go:180              cgocallbackg()
    runtime asm_amd64.s:716             cgocallback_gofunc(0x465463, 0x748810, #21)
    runtime asm_amd64.s:567             asmcgocall(0x748810, #21)
    runtime cgocall.go:124              cgocall(Pointer(0x748810), Pointer(#5))
    qml-go  ??:0                        _Cfunc_applicationExec()
    qml-go  bridge.go:98                RunArgs([]string(#1 len=1 cap=1), func(#4), 0x0)
    qml-go  bridge.go:58                Run(func(#4), 0x0)
    main    frontend.go:63              initFrontend()
    main    main.go:34                  main()
    runtime proc.go:188                 main()
    runtime asm_amd64.s:1998            goexit()
1: chan receive [Created by qml-go.RunArgs @ bridge.go:97]
    qml-go  bridge.go:133               RunMain(#33)
    qml-go  qmlcommon.go:449            (*Common).On(*Common(#3), string(0x8f1520, len=4), interface{}(0x7bd120))
    main    frontend.go:402             (*frontend).loop.func1(*frontend(0x0), 0x0)
    main    frontend.go:414             (*frontend).loop(*frontend(#9), 0x0, 0x0)
    main    frontend.go:63              (*frontend).(main.loop)-fm(0x0, 0x0)
    qml-go  bridge.go:92                RunArgs.func1([]string(#26 len=859534031712 cap=1), func(#25), 0x2)
1: chan receive [Created by main.initFrontend @ frontend.go:62]
    qml-go  bridge.go:133               RunMain(#34)
    qml-go  bridge.go:222               Changed(interface{}(0x8bbe80), interface{}(0x7970c0))
    main    frontend.go:194             (*frontend).qmlBatchLoop(*frontend(#9))
1: chan receive [Created by notify.newNonrecursiveTree @ tree_nonrecursive.go:30]
    notify  tree_nonrecursive.go:81     (*nonrecursiveTree).internal(#17, #16)
2: chan receive [Created by text.(*SerializedBuffer).init @ serialized_buffer.go:26]
    text    serialized_buffer.go:40     (*SerializedBuffer).worker(*)
1: chan receive [Created by backend.newView @ view.go:87]
    backend view.go:217                 (*View).parsethread(#30)
2: chan receive [Created by notify.(*inotify).lazyinit @ watcher_inotify.go:136]
    notify  watcher_inotify.go:253      (*inotify).send(#23, #27)
2: chan receive [Created by watch.NewWatcher @ watcher.go:59]
    watch   watcher.go:239              (*Watcher).observe(*)
1: chan receive [Created by backend.GetEditor @ editor.go:135]
    backend editor.go:342               (*Editor).inputthread(#6)
2: chan receive [Created by log4go.NewConsoleLogWriter @ termlog.go:19]
    log4go  termlog.go:27               ConsoleLogWriter.run(*, #36, #18)
1: chan receive [Created by notify.newNonrecursiveTree @ tree_nonrecursive.go:29]
    notify  tree_nonrecursive.go:36     (*nonrecursiveTree).dispatch(#17, #15)
1: chan receive [Created by log.NewLogWriter @ logwriter.go:41]
    log     logwriter.go:38             NewLogWriter.func1(#8, #7)
1: select [Created by log4go.NewFileLogWriter @ filelog.go:116]
    log4go  filelog.go:84               NewFileLogWriter.func1(#22)
1: syscall [Created by notify.(*inotify).lazyinit @ watcher_inotify.go:133]
    syscall asm_linux_amd64.s:44        Syscall6(0xe8, 0x12, #28, 0x1, 0xffffffffffffffff, 0, 0, 0, 0, 0)
    syscall zsyscall_linux_amd64.go:365 EpollWait(18, []EpollEvent(#28 len=1 cap=1), 18446744073709551615, 0x0, 0x0, 0x0)
    notify  watcher_inotify.go:193      (*inotify).loop(*inotify(#23), <unknown>(#27))
1: syscall [locked]
    runtime asm_amd64.s:1998            goexit()
```

File: cpp/capi.cpp [hunk 1]
```diff
@@ -667,17 +672,17 @@ int objectIsView(QObject_ *object)
     return dynamic_cast<QQuickView *>(qobject) ? 1 : 0;
 }
 
-error *objectGoAddr(QObject_ *object, GoAddr **addr)
+error *objectGoAddr(QObject_ *object, GoValueRef *addr)
```
Comment (erbridge) [inline cpp/capi.cpp:675]: Is this still an `addr`? Both in terms of the function name and the argument.

**Comment by ricochet1k** [comment]
That `fatal error: concurrent map writes` is an interesting line. I wonder where it came from?

**Comment by erbridge** [comment]
Ok, latest commit fixes that issue.

**Comment by neclepsio** [comment]
Is there a motivation why this pull request is not merged, or it will be sometimes in the future? I'd like to migrate my programs from @SjB branch (which provides Go 1.6 compatibility with no other new feature) to this branch, which seems to me the best in active development.

**Comment by ricochet1k** [comment]
@neclepsio Anything that passes across the boundary to QML will not be garbage collected ever. Something needs to be done about that.

**Comment by neclepsio** [comment]
Ok, but since nobody seems to come out with a better idea, isn't it better to have a documented leak instead of a not working at all?

And isn't it the same as with Go <1.5? I don't think the runtime knows if something passed through the boundary can be collected, so IMHO nothing would change.

File: cpp/capi.cpp [hunk 2]
```diff
@@ -45,17 +45,22 @@ void panicf(const char *format, ...)
...
-    static int argc = 1;
-    new QApplication(argc, argv);
+    // if empty is {0} then WM_CLASS is never set due to some crappy logic
+    // in QXcbIntegration::wmClass()
+    static char empty[4] = {'q', 'm', 'l', 0};
+    static char *argv_[] = {empty, 0};
+    static int argc_ = 1;
+    // and, if we use argc and argv, it crashes about
+    // 3/4 of the time with std::bad_alloc()
+    new QApplication(argc_, argv_);
```
Comment (erbridge) [inline cpp/capi.cpp:57]: I bet these were fun to work out...

File: cpp/capi.cpp [hunk 3]
```diff
@@ -228,8 +233,8 @@ void guiappSetWindowIcon(const char *path, int pathLen)
 QQmlComponent_ *newComponent(QQmlEngine_ *engine, QObject_ *parent)
 {
     QQmlEngine *qengine = reinterpret_cast<QQmlEngine *>(engine);
-    //QObject *qparent = reinterpret_cast<QObject *>(parent);
```
Comment (erbridge) [inline cpp/capi.cpp:231]: Any idea why this was commented out before?

File: gomodelindex.go [hunk 4]
```diff
@@ -45,50 +45,50 @@ type ModelIndex interface {
...
 }
 
 func (i *qModelIndex) InternalPointer() uintptr {
-	return uintptr(C.modelIndexInternalPointer(unsafe.Pointer(i.ptr)))
+	return uintptr(C.modelIndexInternalPointer(i.ptr))
 }
 
 func (i *qModelIndex) IsValid() bool {
-	if i == nil || i.ptr == 0 {
+	if i == nil || uintptr(i.ptr) == 0 {
```
Comment (erbridge) [inline gomodelindex.go]: Is this change necessary?

File: qmlcommon.go [hunk 5]
```diff
@@ -437,10 +437,14 @@ func (obj *Common) On(signal string, function interface{}) {
...
 	RunMain(func() {
-		cerr = C.objectConnect(obj.addr, (*C.char)(csignal), C.int(csignallen), obj.engine.addr, unsafe.Pointer(&function), C.int(funcv.Type().NumIn()))
+		fold := &valueFold{gvalue: &function}
+		funcref := getFoldRef(fold)
+		cerr = C.objectConnect(obj.addr, (*C.char)(csignal), C.int(csignallen), obj.engine.addr, funcref, C.int(funcv.Type().NumIn()))
 		if cerr == nil {
 			connectedFunction[&function] = true
 			stats.connectionsAlive(+1)
+		} else {
+
```
Comment (erbridge) [inline qmlcommon.go]: Don't think we need this.

**Review by erbridge — COMMENTED** [review]
Looks good, barring the comments. Does this resolve all of the pointer issues?

File: cpp/capi.cpp (hunk 2 above)
Comment (ricochet1k) [reply cpp/capi.cpp:57]: Haha yeah, not fun. Lots of Googling, reading code, experimenting, and luck.

File: cpp/capi.cpp (hunk 3 above)
Comment (ricochet1k) [reply cpp/capi.cpp:231]: Not a clue. `parent` was never used anyway, but I thought if the code had it, then it should work. I was experimenting with setting the parent to have it destroyed automatically when the engine is destroyed, but that didn't work.

**Comment by elimisteve** [comment]
Does this now work with go1.6?

At https://github.com/limetext/qml-go/issues/3 it is claimed that @SjB added 1.6 support in this branch: https://github.com/SjB/qml/tree/go1.6-port

Anyone want to report on what's working for them and what's not? Thanks!

**Comment by neclepsio** [comment]
Any update on this? I tried using go16 branch but it doesn't compile for me, and fixing what I can results in crashes. Is it my fault or it is not complete?
